Add endpoints to dump selected expert ids

yuhsuan-t opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

When optimizing the performance of MoE models, understanding the expert id distribution helps us to identify the performance bottlenecks and come up with a plan to fix performance issues.  Such information can be captured in `python/sglang/srt/layers/moe/topk.py`.

## Modifications

- Created a singleton class in `python/sglang/srt/managers/utils.py` to record the layer id, expert id, and the topk id in a data structure.
- Called the singleton class in `python/sglang/srt/models/deepseek_v2.py` to recored the layer id into the data structure. (The layer id recording is optional and can be removed.)
- Called the singleton class in `python/sglang/srt/layers/moe/topk.py` to record the expert id and the topk id into the data structure.
- Added two endpoints in `python/sglang/srt/entrypoints/http_server.py` to dump the information captured.  All the other changes under `python/sglang/srt/managers` are related to the two endpoints added.

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [x] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

yuhsuan-t committed:
Add endpoints to dump selected expert ids

yuhsaun-t commented:
Note that this expert id recording involves copying tensors from gpu to cpu during the model execution so will affect performance significantly.  I can add an argument to turn off the recording by default.

yuhsuan-t committed:
Clean up comments

ch-wan commented:
Hi @yuhsaun-t, thank you for your contribution. Could you please provide more details about the output of `dump_record`? The output format is not clear to me. Also, can we use non-blocking copy ([reference](https://pytorch.org/docs/stable/generated/torch.Tensor.copy_.html)) to address the blocking issue?

yuhsaun-t commented:
> Hi @yuhsaun-t, thank you for your contribution. Could you please provide more details about the output of `dump_record`? The output format is not clear to me. Also, can we use non-blocking copy ([reference](https://pytorch.org/docs/stable/generated/torch.Tensor.copy_.html)) to address the blocking issue?

Hello @ch-wan , sure, here is an attached output json from the dump.  Right now there will be one json file for each gpu rank, but the content is the same.  I can update the code to dump only on rank 0.  Yes, I will update to use the non-blocking copy.
[expert_distribution_1738710857.859208.json](https://github.com/user-attachments/files/19351232/expert_distribution_1738710857.859208.json)

yuhsuan-t committed:
Update to non-blocking tensor copy and set start/stop endpoint

This commit lets users turn recording on/off freely so that
by default it does not affect performance.

yuhsuan-t committed:
Remove logging in deepseek_v2 and simplify the output dump format

yuhsuan-t committed:
Add unittest, add doc, format code

yuhsaun-t commented:
Hello @ch-wan , I have updated the PR based on the reviews, and added unittests and docs.
Now the dumped file looks like this:
```
expert_id,count
33,12
4,16
32,8
24,12
40,13
37,11
11,17
```
Can you review the PR again?  Thanks!

ch-wan commented on python/sglang/srt/managers/utils.py:
```diff
@@ -42,3 +47,51 @@ def validate_input_length(
+        for i in topk_ids_list:
+            self._expert_distribution_record.append(tuple(i))
+
+    def reset(self):
```
do we need to unset `self._record`?

yuhsaun-t replied:
Done

ch-wan commented on python/sglang/srt/managers/utils.py:
```diff
@@ -42,3 +47,51 @@ def validate_input_length(
+        self._record = False
+
+    def record_new_token(self, topk_ids):
+        if not self._record:
```
Throw a warning here because it is an unexpected behavior.

yuhsaun-t replied:
Hello @ch-wan , this is not an unexpected behavior.  The `self._record` variable is used to control whether we want to record the expert ids.  Unless the user wants to record the id and has sent request to the `/start_expert_distribution_record` endpoint, by default the expert ids will not be recorded.  This way, the performance will not be affected by the recording.

ch-wan commented on python/sglang/srt/managers/utils.py:
```diff
@@ -42,3 +47,51 @@ def validate_input_length(
+        self._expert_distribution_record.clear()
+
+    def start_record(self):
+        """Start recording the expert distribution. Reset the recorder and set the recording flag to True."""
```
Check `self._record` and throw a warning here if it is `True` because it is an unexpected behavior.

yuhsaun-t replied:
Done

ch-wan commented on python/sglang/srt/managers/utils.py:
```diff
@@ -42,3 +47,51 @@ def validate_input_length(
+        self._record = True
+
+    def stop_record(self):
+        """Stop recording the expert distribution. Set the recording flag to False."""
```
Check `self._record` and throw a warning here if it is `False` because it is an unexpected behavior.

yuhsaun-t replied:
Done

ch-wan commented on python/sglang/srt/managers/utils.py:
```diff
@@ -42,3 +47,51 @@ def validate_input_length(
+            for expert_idx in token_record:
+                results[expert_idx] += 1
+        with open(f"expert_distribution_{time.time()}.csv", "w") as fd:
+            fd.write("expert_id,count\n")
```
My expected output format is `layer_id,expert_id,count` as different layers' experts are independent. But the current output only stores `expert_id` and `count`. Could you please explain it?

yuhsaun-t replied:
@ch-wan Previously I included layer_id here, but the layer_id information can only be retrieved from the model files so I removed it.  I added the layer id recording in `deepseek_v2.py` previously.  If we want to add it back, we'll need to add it to all the current MoE models, and ask developers to add to all the future MoE models.  I haven't thought of a better way to address the issue so I removed it.  If you think there is a better way, or it's fine to add to the model files, I can add it back.  Here is my original code for recording layer id: https://github.com/sgl-project/sglang/pull/4435/commits/50c0e0847ab835df97add252691f90b4fcd57845#diff-5b9e34dd492bd8a14702a18b594721091092276fad1cf8736fba6ef1f33c1b04R1167

ch-wan replied:
I understand it. Mixing experts' workload distribution across layers has limited effective information. Your previous implementation looks good to me. We can support other models later. In `record_new_token`, we can check if `layer_id` is set or not. If not, we can throw an error and let users know that they are probably using a model that does not support recording experts workload.

yuhsaun-t replied:
Added the layer id recording back.  Please check.  Thanks!

ch-wan commented:
@yuhsaun-t Thank you very much for your great effort! This PR would be very useful for analyzing experts' dynamic workloads during MoE serving. I have added some comments. Could you please take a look?

zhyncs committed:
Merge branch 'main' into yuhsuan-t/expert_id_dump

ch-wan reviewed (changes requested):
Another question comes to my mind when I double-check this PR. When DP is enabled, the recorders from different workers will collect different experts distribution. We may need all-reduce to synchronize their results, and only the master worker can dump results.

yuhsuan-t committed:
Update logging messages in utils.py

yuhsaun-t commented:
> Another question comes to my mind when I double-check this PR. When DP is enabled, the recorders from different workers will collect different experts distribution. We may need all-reduce to synchronize their results, and only the master worker can dump results.

The current implementation makes it so that the server will dump one csv file for one rank.  I think we can keep it this way so that the server does not have to synchronize on the fly to save performance.  We can process the dumped csv files later and aggregate them into one file.  Does that sound good to you?

ch-wan commented on python/sglang/srt/managers/utils.py:
```diff
@@ -42,3 +47,61 @@ def validate_input_length(
+        for token_record in self._expert_distribution_record:
+            for expert_idx in token_record:
+                results[expert_idx] += 1
+        with open(f"expert_distribution_{time.time()}.csv", "w") as fd:
```
Thanks for clarifying my previous questions. As multiple workers may dump the results at the same time, they may write their results to the same file. Can we use worker rank as a signature?

yuhsaun-t replied:
I added the rank id to the file name.  Please check.  Thanks!

yuhsuan-t committed:
Add layer id recording back and record rank id

ch-wan commented:
THank you for your great effort! I have approved the change.

ch-wan committed:
Merge branch 'main' into yuhsuan-t/expert_id_dump

yuhsuan-t committed:
Code formatting

ch-wan committed:
fix test

ch-wan committed:
Merge branch 'main' into yuhsuan-t/expert_id_dump

ch-wan committed:
fix test

ch-wan committed:
typo

ch-wan committed:
Merge branch 'main' into yuhsuan-t/expert_id_dump

ch-wan committed:
format

ch-wan committed:
Merge branch 'main' into yuhsuan-t/expert_id_dump

ch-wan commented:
@yuhsaun-t I fixed the issues in your test, but confronted lint issues. Could you please fix it on your side? Thx.

zhyncs committed:
upd

zhyncs committed:
Merge branch 'main' into yuhsuan-t/expert_id_dump

ch-wan commented:
Hi! @yuhsuan-t, could you please join our Slack channel and check your message? Thanks.
